### DISCUSSION TITLE
otPlatAlarm Rollover Cases

### DISCUSSION THREAD

**Discussion by mmb-davidsmith** [openthread/openthread#9065] [Q&A] [answered]
Hi,
I'm looking into our implementation of the otPlatAlarm group of platform support functions. In the documentation for the start functions I see that it says 
> For both aT0 and aDt, the platform MUST support all values in [0, 2^32-1].

As written, I would interpret this to mean that it needs to handle any rollover cases that may be encountered (eg. aT0 = 2^32 - 1000, aDt = 5000, timer should fire at 2^32 + 4000). This is the current main posix platform implementation:

#
```c++
static bool     sIsMsRunning = false;
static uint32_t sMsAlarm     = 0;

void otPlatAlarmMilliStartAt(otInstance *aInstance, uint32_t aT0, uint32_t aDt)
{
    OT_UNUSED_VARIABLE(aInstance);

    sMsAlarm     = aT0 + aDt;
    sIsMsRunning = true;
}
``` 

I don't seem to see any care taken to detect or handle rollover cases. Is this because it's handled at a higher level? From what I can tell, any time the value of aT0 + aDt rolls over, it will cause the timer to fire during the next call of platformAlarmProcess, and platformAlarmUpdateTimeout will set the timeout values to zero.

**Comment by jwhui** [comment, marked as answer]
In the example you gave above, `sMsAlarm` should be set to `4000`.

Then in `platformAlarmUpdateTimeout()`, the timer should fire when `otPlatAlarmMilliGetNow()` is in the past relative to `sMsAlarm`.

https://github.com/openthread/openthread/blob/d9abe3071c0131a4adb5d7e7451319b735e6d855/src/posix/platform/alarm.cpp#L254-L256

One detail regarding "in the past" is that the comparison is done using serial number math. So maybe the confusion is around how large of a delay does the alarm really need to support. I agree that given the current implementation, the largest delay supported relative tot he current time is really 2**31-1.

Does that help? Or is there something else that you were questioning?

**Reply by mmb-davidsmith** [reply]
I think that I was missing the `int32_t` part of that remaining portion in comparing between different platform implementations. A quick check with the following code satisfies me that this isn't going to roll over.

```c++
int32_t remaining;
uint32_t sMsAlarm = 4000;
uint32_t now = UINT32_MAX - 1000;

remaining = (int32_t)(sMsAlarm - now);
printf("sMsAlarm = %"PRIu32", now = %"PRIu32", remaining = %"PRId32"\n", sMsAlarm, now, remaining); 
```

**Reply by mmb-davidsmith** [reply]
So, as a result, with the fact the posix implementation is working this way, should I submit a PR to update the comments on those function definitions to use 2**31-1?

**Reply by jwhui** [reply]
Feel free to propose an update! The intent was really to indicate that the alarm is a 32-bit free-running timer and the platform needs to support that.

**Reply by mmb-davidsmith** [reply]
I've proposed minor changes to the comments in #9067
